pkg/tests/cross-tests/input_check.go

```diff
@@ -0,0 +1,121 @@
+	"github.com/hashicorp/terraform-plugin-go/tftypes"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/pulumi/providertest/providers"
+	"github.com/pulumi/providertest/pulumitest"
+	"github.com/pulumi/providertest/pulumitest/opttest"
+	shimv2 "github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfshim/sdk-v2"
+	"github.com/stretchr/testify/require"
+)
+
+// Adapted from diff_check.go
+type inputTestCase struct {
+	// Schema for the resource under test
+	Resource *schema.Resource
+
+	Config     any
+	ObjectType *tftypes.Object
+}
+
+func assertCtyValEqual(t T, name string, tfVal, pulVal cty.Value) {
+	if !tfVal.RawEquals(pulVal) {
+		t.Logf(name + " not equal!")
+		t.Logf("TF value " + tfVal.GoString())
+		t.Logf("PU value " + pulVal.GoString())
+		t.Fail()
+	}
+}
+
+// Adapted from diff_check.go
+func runCreateInputCheck(t T, tc inputTestCase) {
+	if tc.Resource.CreateContext != nil || tc.Resource.Create != nil {
+		t.Errorf("Create methods should not be set for these tests!")
+	}
+
+	var tfResData, pulResData *schema.ResourceData
+	tc.Resource.CreateContext = func(_ context.Context, rd *schema.ResourceData, meta interface{}) diag.Diagnostics {
+		if tfResData == nil {
+			tfResData = rd
+		} else {
+			pulResData = rd
+		}
+
+		rd.SetId("someid") // CreateContext must pick an ID
+		return make(diag.Diagnostics, 0)
+	}
+	var (
+		providerShortName = "crossprovider"
+		rtype             = "crossprovider_testres"
+		rtok              = "TestRes"
+		rtoken            = providerShortName + ":index:" + rtok
+		providerVer       = "0.0.1"
+	)
+
+	tfwd := t.TempDir()
+
+	tfd := newTfDriver(t, tfwd, providerShortName, rtype, tc.Resource)
+	tfd.writePlanApply(t, tc.Resource.Schema, rtype, "example", tc.Config)
+
+	tfp := &schema.Provider{
+		ResourcesMap: map[string]*schema.Resource{
+			rtype: tc.Resource,
+		},
+	}
+
+	shimProvider := shimv2.NewProvider(tfp, shimv2.WithPlanResourceChange(
+		func(tfResourceType string) bool { return true },
+	))
+
+	pd := &pulumiDriver{
+		name:                providerShortName,
+		version:             providerVer,
+		shimProvider:        shimProvider,
+		pulumiResourceToken: rtoken,
+		tfResourceName:      rtype,
+		objectType:          nil,
+	}
+
+	puwd := t.TempDir()
+	pd.writeYAML(t, puwd, tc.Config)
+
+	pt := pulumitest.NewPulumiTest(t, puwd,
+		opttest.TestInPlace(),
+		opttest.SkipInstall(),
+		opttest.AttachProvider(
+			providerShortName,
+			func(ctx context.Context, pt providers.PulumiTest) (providers.Port, error) {
+				handle, err := pd.startPulumiProvider(ctx)
+				require.NoError(t, err)
+				return providers.Port(handle.Port), nil
+			},
+		),
+		opttest.Env("DISABLE_AUTOMATIC_PLUGIN_ACQUISITION", "true"),
+	)
+
+	pt.Up()
+
+	// compare the two inputs
+	assertCtyValEqual(t, "RawConfig", tfResData.GetRawConfig(), pulResData.GetRawConfig())
+	assertCtyValEqual(t, "RawPlan", tfResData.GetRawPlan(), pulResData.GetRawPlan())
```

Comment:
I would suggest also a TODO here to compare the rest of  `rd *schema.ResourceData`. If you read provider implementation code it rarely reaches to GetRawPlan etc although it does sometimes, but it often reaches for other APIs to extract a field or so. When I was looking at this I found ResourceData incredibly difficult to work with for the purposes of this testing since one needs to know the field keys to pull from before querying it. But perhaps we can construct those from schema, or find some other smart way to compare ResourceData entries.

Reply:
As far as I am aware the .Get methods just grab the data from the config/plan/diff

I'll add a note to verify this.